File: server/use-cases/updateServiceOrder.ts
import { differenceInMinutes } from 'date-fns'
import prisma from "../utils/prisma"

import {
    UpdateServiceOrderParamsType,
} from '../../schemas/preventive'

export async function updateServiceOrder({ id, data: { date, finishTime, startTime, workers } }: UpdateServiceOrderParamsType) {
    try {
        const duration = differenceInMinutes(new Date(finishTime), new Date(startTime))
        await prisma.preventiveOS.update({
            where: {
                id
            },
            data: {
                date: new Date(date),
                duration,
                startTime: new Date(startTime),
                finishTime: new Date(finishTime),
                concluded: true,
                responsible: {
                    set: workers
                }
            }
        })
    } catch (error) {
        throw error
    }
}

File: src-py/use_cases/update_service_order.py
from dateutil import parser

from schemas.preventive import UpdateServiceOrderParams
from utils.date_tools import differenceInMinutes
from infra.prisma import prisma


async def updateServiceOrder(params: dict) -> None:
    validated = UpdateServiceOrderParams.model_validate(params)
    data = validated.data

    duration = differenceInMinutes(data.finishTime, data.startTime)

    await prisma.preventiveos.update(
        where={"id": validated.id},
        data={
            "date": parser.parse(data.date),
            "duration": duration,
            "startTime": parser.parse(data.startTime),
            "finishTime": parser.parse(data.finishTime),
            "concluded": True,
            "responsible": {
                "set": [{"id": worker.id} for worker in data.workers]
            },
        },
    )
